sql/core/src/test/scala/org/apache/spark/sql/sources/TableScanSuite.scala

```diff
@@ -420,4 +434,32 @@ class TableScanSuite extends DataSourceTest with SharedSparkSession {
     val comments = planned.schema.fields.map(_.getComment().getOrElse("NO_COMMENT")).mkString(",")
     assert(comments === "SN,SA,NO_COMMENT")
   }
+
+  test("SPARK-38315: control decoding of datetime as Java 8 classes") {
+    val tableName = "relationProviderWithLegacyTimestamps"
+    def scanTable(): Unit = {
+      withTable (tableName) {
+        sql(
+          s"""
+             |CREATE TABLE $tableName (col TIMESTAMP)
+             |USING org.apache.spark.sql.sources.LegacyTimestampSource
+             """.stripMargin)
+        checkAnswer(
+          spark.table(tableName),
+          Row(java.sql.Timestamp.valueOf("2020-01-01 12:34:56")) :: Nil)
+      }
+    }
+
+    withSQLConf(SQLConf.DATETIME_JAVA8API_ENABLED.key -> "true") {
+      val e = intercept[SparkException] {
+        scanTable()
+      }.getCause
+      assert(e.isInstanceOf[java.lang.RuntimeException])
+      assert(e.getMessage.contains("Error while encoding"))
+
+      withSQLConf(SQLConf.LEGACY_DATETIME_ENCODER_ENABLED.key -> "true") {
```

Comment:
Could you add a comment explaining that this test case is supposed to pass and should not throw any exceptions?